File: apps/front/src/models/order.ts
import { Currency } from '@/lib/currency'

export enum PaymentMethod {
  DonationAlerts = 'donation-alerts',
  Freekassa = 'freekassa',
  EasyDonate = 'easy-donate',
}

export type CreateOrderReq = {
  paymentMethod: PaymentMethod
  products: CreateOrderProductReq[]
}

export type CreateOrderProductReq = {
  id: string
  profileId: string
}

export type CreateOrderRes = {
  paymentUrl: string
}

export type OrderStatus = 'created' | 'processing' | 'completed' | 'failed'

export type OrderItem = {
  id: string
  productId: string
  profileId: string
  seasonId: string
  quantity: number
  amounts: OrderAmount[]
}

export type OrderAmount = {
  currency: Currency
  amount: number
}

export type Order = {
  id: string
  status: OrderStatus
  amounts: OrderAmount[]
  items: OrderItem[]
  createdAt: string
}

export type PurchasedProduct = {
  productId: string
  profileId: string
  seasonId: string
}

File: apps/api/internal/domain/order.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

type OrderStatus string

const (
	OrderStatusCreated    OrderStatus = "created"
	OrderStatusProcessing OrderStatus = "processing"
	OrderStatusCompleted  OrderStatus = "completed"
	OrderStatusFailed     OrderStatus = "failed"
)

type PaymentMethod string

const (
	PaymentMethodEasyDonate PaymentMethod = "easy-donate"
)

type OrderAmounts struct {
	Currency Currency `json:"c"`
	Amount   float64  `json:"a"`
}

type Order struct {
	ID            uuid.UUID
	UserID        uuid.UUID
	Amounts       []*OrderAmounts
	Status        OrderStatus
	PaymentMethod PaymentMethod
	ExternalID    string
	CreatedAt     time.Time
	CreatedBy     *uuid.UUID
	UpdatedAt     time.Time
	UpdatedBy     *uuid.UUID
	// relations
	Items []*OrderItem
}

type OrderItem struct {
	ID        uuid.UUID
	OrderID   uuid.UUID
	ProductID uuid.UUID
	ProfileID uuid.UUID
	SeasonID  uuid.UUID
	Amounts   []*OrderAmounts
	Quantity  int
	// relations
	Order   *Order
	Product *Product
	Profile *Profile
	Season  *Season
}

// OrderStatuses lists every order status in the order an order goes through them.
var OrderStatuses = []OrderStatus{OrderStatusCreated, OrderStatusProcessing, OrderStatusCompleted, OrderStatusFailed}

// OrderFilter narrows the admin order list. An empty status keeps every status.
// Search matches the order ID, the EasyDonate payment ID or a recipient username, all by prefix.
type OrderFilter struct {
	Status OrderStatus
	Search string
}
